cpp/include/cudf/fixed_point/fixed_point.hpp

```diff
@@ -0,0 +1,295 @@
+}
+
+template <typename Rep, typename T>
+auto division_overflow(T lhs, T rhs) -> bool {
+    return lhs == std::numeric_limits<Rep>::min() && rhs == -1;
+}
+
+template <typename Rep, typename T>
+auto multiplication_overflow(T lhs, T rhs) -> bool {
+    auto const min = std::numeric_limits<Rep>::min();
+    auto const max = std::numeric_limits<Rep>::max();
+    if      (rhs >  0) return lhs > max / rhs || lhs < min / rhs;
+    else if (rhs < -1) return lhs > min / rhs || lhs < max / rhs;
+    else               return rhs == -1 && lhs == min;
+}
+
+// PLUS Operation
+template<typename Rep1, Radix Rad1,
+         typename Rep2, Radix Rad2>
+fixed_point<Rep1, Rad1> operator+(fixed_point<Rep1, Rad1> const& lhs,
+                                  fixed_point<Rep2, Rad2> const& rhs) {
+
+    static_assert(std::is_same<Rep1, Rep2>::value, "Represenation types should be the same");
+    static_assert(Rad1 == Rad2,                    "Radix types should be the same");
+
+    auto const rhsv  = lhs._scale > rhs._scale ? detail::shift<Rad1>(rhs._value, scale_type{lhs._scale - rhs._scale}) : rhs._value;
+    auto const lhsv  = lhs._scale < rhs._scale ? detail::shift<Rad1>(lhs._value, scale_type{rhs._scale - lhs._scale}) : lhs._value;
+    auto const scale = lhs._scale > rhs._scale ? lhs._scale : rhs._scale;
+
+    #if defined(__CUDACC_DEBUG__)
+
+    assert(!addition_overflow<Rep1>(lhsv, rhsv) &&
+        "fixed_point overflow of underlying represenation type " + print_rep<Rep1>());
+
+    #endif
+
+    return fixed_point<Rep1, Rad1>{scaled_integer<Rep1>(lhsv + rhsv, scale)};
+}
+
+// MINUS Operation
+template<typename Rep1, Radix Rad1,
+         typename Rep2, Radix Rad2>
+fixed_point<Rep1, Rad1> operator-(fixed_point<Rep1, Rad1> const& lhs,
+                                  fixed_point<Rep2, Rad2> const& rhs) {
+
+    static_assert(std::is_same<Rep1, Rep2>::value, "Represenation types should be the same");
+    static_assert(Rad1 == Rad2,                    "Radix types should be the same");
+
+    auto const rhsv  = lhs._scale > rhs._scale ? detail::shift<Rad1>(rhs._value, scale_type{lhs._scale - rhs._scale}) : rhs._value;
+    auto const lhsv  = lhs._scale < rhs._scale ? detail::shift<Rad1>(lhs._value, scale_type{rhs._scale - lhs._scale}) : lhs._value;
+    auto const scale = lhs._scale > rhs._scale ? lhs._scale : rhs._scale;
+
+    #if defined(__CUDACC_DEBUG__)
+
+    assert(!subtraction_overflow<Rep1>(lhsv, rhsv) &&
+        "fixed_point overflow of underlying represenation type " + print_rep<Rep1>());
+
+    #endif
+
+    return fixed_point<Rep1, Rad1>{scaled_integer<Rep1>(lhsv - rhsv, scale)};
+}
+
+// MULTIPLIES Operation
+template<typename Rep1, Radix Rad1,
+         typename Rep2, Radix Rad2>
+fixed_point<Rep1, Rad1> operator*(fixed_point<Rep1, Rad1> const& lhs,
+                                  fixed_point<Rep2, Rad2> const& rhs) {
+
+    static_assert(std::is_same<Rep1, Rep2>::value, "Represenation types should be the same");
```

Comment:
Instead of needing all these static asserts, just fix the `Rep` and `Rad` to be the same for `lhs/rhs`.

```
template <typename Rep, Radix Rad>
fixed_point<Rep, Rad> operator*(fixed_point<Rep, Rad> const& lhs, fixed_point<Rep, Rad> const& rhs);

Reply:
I had originally thought of that but thought this would be clearer from a "compile time warning" point of view. However, the way it is being used (aka through decimal32) this situation probably won't be run into to often, so I agree - this is probably unnecessary. Will do.

Reply:
Done